### PR TITLE
Add checklist for Design Guidelines to doc/DESIGN.md - merge doc/todo…

### PR DESCRIPTION
**Pull request by lawli3t**
…/API into doc/DESIGN.md

## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: doc/DESIGN.md [hunk 1]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+On potential changes of the API/ABI as detected by
+[`elektra-icheck`](https://build.libelektra.org/job/elektra-icheck),
+please make sure the API has been reviewed according to the
+following 2 checklists:
+
+## Checklist for overall API
+
+### Naming
+
+- [ ] Inconsistent naming of functions or variables
```
Comment (markus2330) [inline doc/DESIGN.md:298]: What do you mean with variables? We have enums, macros, typedefs and functions.

File: doc/DESIGN.md [hunk 2]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+### Consistency
+
+- [ ] Consistent naming schemes
```
Comment (markus2330) [inline doc/DESIGN.md:302]: How does this differ to the point above?

File: doc/DESIGN.md [hunk 3]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Similar things are named similarly
+- [ ] Different things are named differently
+- [ ] The order of arguments should be consistent across similar functions
+- [ ] Code is formatted according to the [style guidelines](/doc/CODING.md#coding-style)
```
Comment (markus2330) [inline doc/DESIGN.md:306]: Code should be irrelevant here, it is only about API?

File: doc/DESIGN.md [hunk 4]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+### Structural Clarity
+
+- [ ] Functions with similar functionality are grouped into the same namespace
```
Comment (markus2330) [inline doc/DESIGN.md:310]: We do not have namespaces per se but we prefix function names.

File: doc/DESIGN.md [hunk 5]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+### Compatibility
+
+- [ ] All Bindings have been updated to reflect the new API and work properly
+
+### Extensibility
+
+- [ ] New API is easily extensible with additional functionality
+- [ ] Components do not depend too heavily on each other
```
Comment (markus2330) [inline doc/DESIGN.md:319]: Can you give an example what you mean here?

File: doc/DESIGN.md [hunk 6]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+- [ ] Functions with similar functionality are grouped into the same namespace
+
+### Compatibility
+
+- [ ] All Bindings have been updated to reflect the new API and work properly
+
+### Extensibility
+
+- [ ] New API is easily extensible with additional functionality
```
Comment (markus2330) [inline doc/DESIGN.md:318]: Unclear how this can be checked.

File: doc/DESIGN.md [hunk 7]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Components do not depend too heavily on each other
+
+## Checklist for each function
+
+### Documentation
+
+- [ ] Doxygen Documentation is complete
+      (covers all parameters, brief/short summary, examples)
```
Comment (markus2330) [inline doc/DESIGN.md:326]: Give a check list of what a complete Doxygen docu is:

- [ ] return/retval
- [ ] see
- [ ] pre/post/invariant
- [ ] since (which is not done at the moment for most functions)

File: doc/DESIGN.md [hunk 8]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Change is mentioned in the release notes
```
Comment (markus2330) [inline doc/DESIGN.md:327]:
```suggestion
- [ ] Change is mentioned in the Compatibility section of the release notes
```

File: doc/DESIGN.md [hunk 9]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Inconsistencies between documentation and code
+- [ ] Inconsistencies between documentation and tests
+- [ ] Proper Documentation of all side effects
```
Comment (markus2330) [inline doc/DESIGN.md:330]: What do you mean here? Actually, Elektra's functions should not have side effects, they should only update objects.

File: doc/DESIGN.md [hunk 10]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+- [ ] Abbreviations should be avoided in function / parameter names
+- [ ] Function / parameter names should neither be too long, nor too short
+- [ ] Function name should be clear and unambiguous
+
+### Compatibility
+
+- [ ] ABI/API forward-compatible (breaking backwards-compatibility
+      to add new symbols is fine)
+- [ ] #ifdef present for experimental features
```
Comment (markus2330) [inline doc/DESIGN.md:345]: At the moment we do not have a proper process how to introduce experimental features. The idea of libproposal did not work well. Macros we did not try yet.

File: doc/DESIGN.md [hunk 11]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+### Parameter & Return Types
+
+- [ ] Returning a not specific enough type
+- [ ] Requiring a not liberal enough parameter
+- [ ] should not have multiple parameters of the same type
```
Comment (markus2330) [inline doc/DESIGN.md:351]:
```suggestion
- [ ] Should not have multiple parameters of the same type
```

How you want to do a copy or similar then?

File: doc/DESIGN.md [hunk 12]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Functions should use Constant Types instead of Boolean Types wherever
```
Comment (markus2330) [inline doc/DESIGN.md:352]:
```suggestion
- [ ] Functions should use constant types instead of boolean types
```

What is a constant type?

File: doc/DESIGN.md [hunk 13]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+      sensible
+- [ ] Wherever possible, function parameters should be `const`
+- [ ] Functions should not have a long list of parameters (>8)
```
Comment (markus2330) [inline doc/DESIGN.md:355]: At the moment we do not even have 4.

File: doc/DESIGN.md [hunk 14]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+### Error Handling
+
+- [ ] When an error occurs, a clear error message should be provided
+- [ ] Errors of the same type should emit the same error message
+- [ ] The error message informs the user of possible causes for the problem
+- [ ] All possible error messages are documented
```
Comment (markus2330) [inline doc/DESIGN.md:362]: This is not needed, the function should only say which error categories are possible, and if API users can have a useful reaction to them.

File: doc/DESIGN.md [hunk 15]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] All possible errors states lead to an error
+- [ ] Proper error codes are chosen
+
+### Structural Clarity
+
+- [ ] Function does exactly one thing
```
Comment (markus2330) [inline doc/DESIGN.md:368]: It would be good if you consistently use must/should/may as in requirements.

File: doc/DESIGN.md [hunk 16]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Function should have no side effects
+
+### Memory Management
+
+- [ ] Memory Management should be handled by the function wherever possible
+- [ ] Functions should not cause memory-leaks
+- [ ] Every Buffer should be accompanied by a max size limit
```
Comment (markus2330) [inline doc/DESIGN.md:375]: I think this point is too obvious and does not need to be checked.

File: doc/DESIGN.md [hunk 17]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+- [ ] Functions who require a large amount of memory to be allocated,
+      state so in their documentation
+
+### Extensibility
+
+- [ ] Function is easily extensible with additional functionality
```
Comment (markus2330) [inline doc/DESIGN.md:381]: I think this should be more concrete: e.g. customizable actions via flags if useful.

File: doc/DESIGN.md [hunk 18]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+### Documentation
+
+- [ ] Doxygen Documentation is complete
+      (covers all parameters, brief/short summary, examples)
+- [ ] Change is mentioned in the release notes
+- [ ] Inconsistencies between documentation and code
+- [ ] Inconsistencies between documentation and tests
+- [ ] Proper Documentation of all side effects
+- [ ] Proper Documentation of thread-safety of function
+- [ ] [Symbol versioning](/doc/dev/symbol-versioning.md)
```
Comment (markus2330) [inline doc/DESIGN.md:332]: This is not docu.

File: doc/DESIGN.md [hunk 19]
```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
...
+
+### Extensibility
+
+- [ ] Function is easily extensible with additional functionality
+
+### Tests
+
+- [ ] Added functions are fully covered by tests
+- [ ] Tests cover edge-cases
+- [ ] Tests cover error-cases and check for proper error messages
```
Comment (markus2330) [inline doc/DESIGN.md:387]: Again, often it is enough to check for error categories.

**Review by markus2330 — COMMENTED** [review]
Thank you, this is a very extensive checklist.

File: doc/DESIGN.md (hunk 6 above)
Comment (lawli3t) [reply doc/DESIGN.md:318]: Yes this was unclear to me as well. You mentioned it in our email conversation - so I put it up here for visibility. I haven't really found any good heuristics for this yet.

File: doc/DESIGN.md (hunk 5 above)
Comment (lawli3t) [reply doc/DESIGN.md:319]: This is basically the loosely-coupled principle from SOLID. I think I still have a problem of distinguishing some guidelines that are only applicable to specific parts of the codebase. (Core, Bindings,...)

File: doc/DESIGN.md (hunk 10 above)
Comment (lawli3t) [reply doc/DESIGN.md:345]: okay, I only included this because this was part of the deleted API file in the todo folder. Should I just remove and forego this until there is a proper solution in place?

File: doc/DESIGN.md (hunk 13 above)
Comment (lawli3t) [reply doc/DESIGN.md:355]: Nice, should I reduce the number included or remove the whole line altogether? This was more or less arbitrarily chosen and can definitely be change to accomodate the project.

File: doc/DESIGN.md (hunk 15 above)
Comment (lawli3t) [reply doc/DESIGN.md:368]: That is a very good idea, I assume you refer to the definitions in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)?

Comment (markus2330) [reply doc/DESIGN.md]: Yes, https://www.ietf.org/rfc/rfc2119.txt is a good starting point. We might want to allow additional exceptions though.

File: doc/DESIGN.md (hunk 5 above)
Comment (markus2330) [reply doc/DESIGN.md:315]: Ahh, ok. Yes, between the various libs it makes sense to check this. Within the core, however, it is quite limited what can be done.

File: doc/DESIGN.md (hunk 6 above)
Comment (markus2330) [reply doc/DESIGN.md:314]: Yes, being futureproof is a difficult topic.

One part of it is to make sure that we are generic, e.g. general flags instead of specific locking flags. But being too generic also is problematic as people might not think of that they can lock a key via setting flags (or remove a key via lookup).

One other part of it is to have as little assumptions as possible on what users might want to do.

Did you find some papers about the topic?

File: doc/DESIGN.md (hunk 10 above)
Comment (markus2330) [reply doc/DESIGN.md]: Remove it. With this extensive checklist we actually already make quite sure that new additions to the API do not have obvious problems.

File: doc/DESIGN.md (hunk 13 above)
Comment (markus2330) [reply doc/DESIGN.md]: This is basically the counterpart of being generic. With more generic functions (e.g. flags to modify behavior) more parameters are added. Maybe we simply add here in the checklist that these trade-offs should be considered, without giving numbers or judgements how the trade-off should be chosen as this is already described in doc/GOALS.md

File: doc/DESIGN.md (hunk 6 above)
Comment (lawli3t) [reply doc/DESIGN.md:314]: I haven't really found anything that can be easily distilled into a checklist. I have been doing further research into this area, so I hope I can can come up with something soon.
